File: fhir-bridge/node/src/fhirToIntake.ts
/**
 * FHIR R5 Bundle -> IntakeFamily.
 *
 * Reads FamilyMemberHistory resources, maps their relationship code to
 * an Evagene relative_type, and emits a flat proband+relatives shape
 * the IntakeService can create in Evagene.
 *
 * Pure transformation: unknown codes produce warnings, never exceptions.
 */

import type {
  FhirBundle,
  FhirBundleEntry,
  FhirCodeableConcept,
  FhirFamilyMemberHistory,
} from './fhirTypes.js';
import type {
  BiologicalSex,
  IntakeFamily,
  Relative,
  RelativeType,
} from './intakeFamily.js';
import { MappingError } from './pedigreeToFhir.js';
import { FHIR_V3_ROLE_CODE_SYSTEM, mapToEvagene } from './relationMap.js';

export interface FhirToIntakeResult {
  readonly family: IntakeFamily;
  readonly warnings: readonly string[];
}

export interface FhirToIntakeOptions {
  readonly patientId: string;
  readonly probandDisplayName?: string;
}

export function fhirBundleToIntakeFamily(
  bundle: unknown,
  options: FhirToIntakeOptions,
): FhirToIntakeResult {
  const resources = readFamilyMemberHistoryResources(bundle);
  const warnings: string[] = [];
  const relatives: Relative[] = [];

  for (const resource of resources) {
    const mapped = mapResource(resource, warnings);
    if (mapped !== undefined) {
      relatives.push(mapped);
    }
  }

  const probandName = (options.probandDisplayName ?? '').trim() || `Patient ${options.patientId}`;
  return {
    family: {
      pedigreeDisplayName: `${probandName}'s family (from FHIR)`,
      proband: { displayName: probandName, biologicalSex: 'unknown' },
      relatives,
    },
    warnings,
  };
}

function mapResource(
  resource: FhirFamilyMemberHistory,
  warnings: string[],
): Relative | undefined {
  const fhirCode = extractV3Code(resource.relationship);
  if (fhirCode === undefined) {
    warnings.push(
      `skipped FamilyMemberHistory ${resource.id ?? '(no id)'}: no v3-RoleCode relationship coding.`,
    );
    return undefined;
  }
  const evageneType = mapToEvagene(fhirCode);
  if (evageneType === undefined) {
    warnings.push(
      `skipped FamilyMemberHistory ${resource.id ?? '(no id)'}: relationship code '${fhirCode}' is not supported.`,
    );
    return undefined;
  }
  return buildRelative(resource, evageneType);
}

function buildRelative(resource: FhirFamilyMemberHistory, relativeType: RelativeType): Relative {
  const displayName = resource.name ?? resource.id ?? 'Unnamed relative';
  const sex = extractSex(resource.sex);
  const year = extractYearOfBirth(resource.bornDate);
  return year === undefined
    ? { relativeType, displayName, biologicalSex: sex }
    : { relativeType, displayName, biologicalSex: sex, yearOfBirth: year };
}

function readFamilyMemberHistoryResources(raw: unknown): readonly FhirFamilyMemberHistory[] {
  if (typeof raw !== 'object' || raw === null) {
    throw new MappingError('FHIR response is not a Bundle with entries.');
  }
  const candidate = raw as Partial<FhirBundle>;
  const entries: readonly FhirBundleEntry[] | undefined = candidate.entry;
  if (candidate.resourceType !== 'Bundle' || entries === undefined) {
    throw new MappingError('FHIR response is not a Bundle with entries.');
  }
  const resources: FhirFamilyMemberHistory[] = [];
  for (const entry of entries) {
    const resource = entry.resource;
    if (resource?.resourceType === 'FamilyMemberHistory') {
      resources.push(resource);
    }
  }
  return resources;
}

function extractV3Code(concept: FhirCodeableConcept | undefined): string | undefined {
  for (const coding of concept?.coding ?? []) {
    if (coding.system === FHIR_V3_ROLE_CODE_SYSTEM) {
      return coding.code;
    }
  }
  return undefined;
}

function extractSex(concept: FhirCodeableConcept | undefined): BiologicalSex {
  for (const coding of concept?.coding ?? []) {
    const code = coding.code;
    if (code === 'female' || code === 'male') {
      return code;
    }
  }
  return 'unknown';
}

function extractYearOfBirth(bornDate: string | undefined): number | undefined {
  if (bornDate === undefined) {
    return undefined;
  }
  const match = /^(\d{4})/.exec(bornDate);
  const captured = match?.[1];
  if (captured === undefined) {
    return undefined;
  }
  return Number.parseInt(captured, 10);
}

File: fhir-bridge/dotnet/src/FhirBridge/FhirToIntake.cs
using System.Text.Json;

namespace FhirBridge;

public sealed record FhirToIntakeOptions(string PatientId, string? ProbandDisplayName = null);

public sealed record FhirToIntakeResult(IntakeFamily Family, IReadOnlyList<string> Warnings);

public static class FhirToIntake
{
    public static FhirToIntakeResult ToIntakeFamily(JsonElement bundle, FhirToIntakeOptions options)
    {
        if (bundle.ValueKind != JsonValueKind.Object ||
            !TryReadResourceType(bundle, out var resourceType) ||
            resourceType != "Bundle")
        {
            throw new MappingException("FHIR response is not a Bundle with entries.");
        }
        if (!bundle.TryGetProperty("entry", out var entries) || entries.ValueKind != JsonValueKind.Array)
        {
            throw new MappingException("FHIR response is not a Bundle with entries.");
        }

        var warnings = new List<string>();
        var relatives = new List<Relative>();
        foreach (var entry in entries.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Object ||
                !entry.TryGetProperty("resource", out var resource) ||
                resource.ValueKind != JsonValueKind.Object ||
                !TryReadResourceType(resource, out var kind) ||
                kind != "FamilyMemberHistory")
            {
                continue;
            }
            var mapped = MapResource(resource, warnings);
            if (mapped is not null)
            {
                relatives.Add(mapped);
            }
        }

        var probandName = (options.ProbandDisplayName ?? string.Empty).Trim();
        if (probandName.Length == 0)
        {
            probandName = $"Patient {options.PatientId}";
        }

        var family = new IntakeFamily(
            PedigreeDisplayName: $"{probandName}'s family (from FHIR)",
            Proband: new Proband(probandName, BiologicalSex.Unknown),
            Relatives: relatives);
        return new FhirToIntakeResult(family, warnings);
    }

    private static Relative? MapResource(JsonElement resource, List<string> warnings)
    {
        var id = TryReadString(resource, "id") ?? "(no id)";
        var fhirCode = ExtractV3Code(resource);
        if (fhirCode is null)
        {
            warnings.Add($"skipped FamilyMemberHistory {id}: no v3-RoleCode relationship coding.");
            return null;
        }
        var evageneType = RelationMap.ToEvagene(fhirCode);
        if (evageneType is null)
        {
            warnings.Add($"skipped FamilyMemberHistory {id}: relationship code '{fhirCode}' is not supported.");
            return null;
        }

        var displayName = TryReadString(resource, "name") ?? id;
        var sex = ExtractSex(resource);
        var year = ExtractYearOfBirth(resource);
        return new Relative(evageneType.Value, displayName, sex, year);
    }

    private static string? ExtractV3Code(JsonElement resource)
    {
        if (!resource.TryGetProperty("relationship", out var relationship) ||
            relationship.ValueKind != JsonValueKind.Object ||
            !relationship.TryGetProperty("coding", out var coding) ||
            coding.ValueKind != JsonValueKind.Array)
        {
            return null;
        }
        foreach (var entry in coding.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Object) continue;
            if (TryReadString(entry, "system") == RelationMap.V3RoleCodeSystem)
            {
                return TryReadString(entry, "code");
            }
        }
        return null;
    }

    private static BiologicalSex ExtractSex(JsonElement resource)
    {
        if (!resource.TryGetProperty("sex", out var sex) ||
            sex.ValueKind != JsonValueKind.Object ||
            !sex.TryGetProperty("coding", out var coding) ||
            coding.ValueKind != JsonValueKind.Array)
        {
            return BiologicalSex.Unknown;
        }
        foreach (var entry in coding.EnumerateArray())
        {
            if (entry.ValueKind != JsonValueKind.Object) continue;
            var code = TryReadString(entry, "code");
            if (code == "female") return BiologicalSex.Female;
            if (code == "male") return BiologicalSex.Male;
        }
        return BiologicalSex.Unknown;
    }

    private static int? ExtractYearOfBirth(JsonElement resource)
    {
        var bornDate = TryReadString(resource, "bornDate");
        if (bornDate is null || bornDate.Length < 4)
        {
            return null;
        }
        return int.TryParse(bornDate.AsSpan(0, 4), out var year) ? year : null;
    }

    private static bool TryReadResourceType(JsonElement element, out string value)
    {
        if (element.TryGetProperty("resourceType", out var v) && v.ValueKind == JsonValueKind.String)
        {
            value = v.GetString() ?? string.Empty;
            return true;
        }
        value = string.Empty;
        return false;
    }

    private static string? TryReadString(JsonElement element, string key)
    {
        if (element.TryGetProperty(key, out var value) && value.ValueKind == JsonValueKind.String)
        {
            return value.GetString();
        }
        return null;
    }
}
